x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java

```diff
@@ -352,76 +358,78 @@ public String toString() {
-                                        "failed to warm cache for [{}] part [{}/{}]",
-                                        fileName,
-                                        part,
-                                        numberOfParts
-                                    ),
-                                    e
-                                );
-                            }
-
-                            @Override
-                            public void onAfter() {
-                                if (countDown.countDown()) {
-                                    IOUtils.closeWhileHandlingException(input);
-                                }
-                            }
-                        });
-                    }
-                } catch (IOException e) {
-                    logger.trace(() -> new ParameterizedMessage("failed to warm cache for [{}]", fileName), e);
-                }
+            logger.debug("{} warming shard cache for [{}] files", shardId, queue.size());
+
+            // Start as many workers as fit into the searchable snapshot pool at once at the most
+            final int workers = Math.min(threadPool.info(SEARCHABLE_SNAPSHOTS_THREAD_POOL_NAME).getMax(), queue.size());
+            for (int i = 0; i < workers; ++i) {
+                prewarmNextFile(executor, queue);
             }
         }
     }
 
+    private void prewarmNextFile(final Executor executor, final BlockingQueue<BlobStoreIndexShardSnapshot.FileInfo> queue) {
+        executor.execute(new AbstractRunnable() {
+            @Override
+            protected void doRun() throws Exception {
+                final BlobStoreIndexShardSnapshot.FileInfo file = queue.poll(0L, TimeUnit.MILLISECONDS);
+                if (file == null) {
+                    return;
+                }
+
+                final IndexInput input = openInput(file.physicalName(), CachedBlobContainerIndexInput.CACHE_WARMING_CONTEXT);
+                assert input instanceof CachedBlobContainerIndexInput : "expected cached index input but got " + input.getClass();
+
+                final int numberOfParts = Math.toIntExact(file.numberOfParts());
+                final GroupedActionListener<Void> listener = new GroupedActionListener<>(
+                    ActionListener.runAfter(
+                        ActionListener.wrap(() -> IOUtils.closeWhileHandlingException(input)),
+                        () -> prewarmNextFile(executor, queue)
+                    ),
+                    numberOfParts
+                );
+
+                // if the file to prewarm is composed of a single part then it is prewarmed using the current thread
+                final Executor warmExecutor = (numberOfParts > 1) ? executor : EsExecutors.newDirectExecutorService();
+
+                for (int p = 0; p < numberOfParts; p++) {
+                    final int part = p;
+                    warmExecutor.execute(ActionRunnable.run(listener, () -> {
```

Comment:
Not sure it's a big deal but I guess we could always use the current thread for one part:

```suggestion
                     final Executor warmExecutor = (p < numberOfParts - 1) ? executor : EsExecutors.newDirectExecutorService();
                    warmExecutor.execute(ActionRunnable.run(listener, () -> {
```